src/net.cpp

```diff
@@ -2190,7 +2191,41 @@ NodeId CConnman::GetNewNodeId()
     return nLastNodeId.fetch_add(1, std::memory_order_relaxed);
 }
 
-bool CConnman::Start(CScheduler& scheduler, std::string& strNodeError, Options connOptions)
+
+bool CConnman::Bind(const CService &addr, unsigned int flags) {
+    if (!(flags & BF_EXPLICIT) && IsLimited(addr))
+        return false;
+    std::string strError;
+    if (!BindListenPort(addr, strError, (flags & BF_WHITELIST) != 0)) {
+        if ((flags & BF_REPORT_ERROR) && clientInterface) {
+            clientInterface->ThreadSafeMessageBox(strError, "", CClientUIInterface::MSG_ERROR);
+        }
+        return false;
+    }
+    return true;
+}
+
+bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<CService>& whiteBinds) {
+    if (fListen) {
+        bool fBound = false;
+        for (const auto& addrBind : binds) {
```

Comment:
> Mm, this moves the logic away from the caller and into CConnman, which is what we're trying to avoid by adding the heaps of options.
> I think I'd prefer to see flags passed in along with CServices, so that CConnman can just do what it's told. So either a tiny struct, or just a pair.

The logic would be in `CConnman` all the same, just the data structure you pass the info with is changed from two vectors to one vector of pairs, which is equivalent. The code looks more complicated from what I can tell. Or did I misunderstand your request?

I added this in a separate commit to see the diff, I drop it again if needed.

> Then, CConnman will try to bind the BF_EXPLICIT addresses, and if it fails, tries to bind the rest.

At the moment, the rest is only attempted if no explicit binds are specified. If they fail, all fails, which makes sense to me.

> Also, in that case, the last failed bind wouldn't need to have BF_REPORT_ERROR set, we'd just notice the false return and act on it accordingly.

Act how, apart from reporting the error (which this flag does)?